src/drizzleReactHooks.js

```diff
@@ -0,0 +1,203 @@
+    argsRef.current = args
+    const newState = mapStateRef.current(drizzle.store.getState())
+    if (!shallowequal(stateRef.current, newState)) {
+      stateRef.current = newState
+      setState(newState)
+    }
+  }
+  useEffect(
+    () => {
+      const debouncedHandler = debounce(() => {
+        const newState = mapStateRef.current(drizzle.store.getState())
+        if (!shallowequal(stateRef.current, newState)) {
+          stateRef.current = newState
+          setState(newState)
+        }
+      })
+      const unsubscribe = drizzle.store.subscribe(debouncedHandler)
+      return () => {
+        unsubscribe()
+        debouncedHandler.clear()
+      }
+    },
+    [drizzle.store]
+  )
+  return stateRef.current
+}
+
+export const DrizzleProvider = ({ children, drizzle }) => {
+  const useCacheCall = useCallback(
+    (contractNameOrContractNames, methodNameOrFunction, ...args) => {
+      const isFunction = typeof methodNameOrFunction === 'function'
+      const drizzleState = useDrizzleState(drizzleState => {
+        if (isFunction)
+          return contractNameOrContractNames.reduce((acc, contractName) => {
+            acc[contractName] = drizzleState.contracts[contractName]
+            return acc
+          }, {})
+        else {
+          const cacheKey = drizzle.contracts[
+            contractNameOrContractNames
+          ].methods[methodNameOrFunction].cacheCall(...args)
+          return {
+            value:
+              drizzleState.contracts[contractNameOrContractNames][
+                methodNameOrFunction
+              ][cacheKey] &&
+              drizzleState.contracts[contractNameOrContractNames][
+                methodNameOrFunction
+              ][cacheKey].value
+          }
+        }
+      }, args)
+      return isFunction
+        ? methodNameOrFunction((contractName, methodName, ...args) => {
+            const cacheKey = drizzle.contracts[contractName].methods[
+              methodName
+            ].cacheCall(...args)
+            return (
+              drizzleState[contractName][methodName][cacheKey] &&
+              drizzleState[contractName][methodName][cacheKey].value
+            )
+          })
+        : drizzleState.value
+    },
+    [drizzle.contracts]
+  )
+  const useCacheSend = useCallback(
+    (contractName, methodName) => {
+      const drizzleState = useDrizzleState(drizzleState => ({
+        transactionStack: drizzleState.transactionStack,
+        transactions: drizzleState.transactions
+      }))
+      const [stackIDs, setStackIDs] = useState([])
+      const transactions = stackIDs.map(
+        stackID =>
+          drizzleState.transactions[
+            drizzleState.transactionStack[stackID] || 'undefined'
+          ]
+      )
+      return {
+        send: (...args) =>
+          setStackIDs(stackIDs => [
+            ...stackIDs,
+            drizzle.contracts[contractName].methods[methodName].cacheSend(
+              ...args
```

Comment:
This nesting is kind of confusing, could you refactor with some intermediate variables?

Reply:
sure